// remult/postgres/postgres-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.preparePostgresQueueStorage = exports.createPostgresDataProvider = exports.createPostgresConnection = exports.PostgresDataProvider = void 0;
var tslib_1 = require("tslib");
var pg_1 = require("pg");
var context_1 = require("../src/context");
var sql_database_1 = require("../src/data-providers/sql-database");
var schema_builder_1 = require("./schema-builder");
var PostgresDataProvider = /** @class */ (function () {
    function PostgresDataProvider(pool) {
        this.pool = pool;
        this.supportsJsonColumnType = true;
    }
    PostgresDataProvider.getDb = function (remult) {
        var sql = sql_database_1.SqlDatabase.getDb(remult);
        var me = sql._getSourceSql();
        if (!me.pool) {
            throw 'the data provider is not a PostgresDataProvider';
        }
        return me.pool;
    };
    PostgresDataProvider.prototype.entityIsUsedForTheFirstTime = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/];
        }); });
    };
    PostgresDataProvider.prototype.getLimitSqlSyntax = function (limit, offset) {
        return ' limit ' + limit + ' offset ' + offset;
    };
    PostgresDataProvider.prototype.createCommand = function () {
        return new PostgresBridgeToSQLCommand(this.pool);
    };
    PostgresDataProvider.prototype.ensureSchema = function (entities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var db, sb;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = new sql_database_1.SqlDatabase(this);
                        sb = new schema_builder_1.PostgresSchemaBuilder(db);
                        return [4 /*yield*/, sb.ensureSchema(entities)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PostgresDataProvider.prototype.transaction = function (action) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var client, err_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.pool.connect()];
                    case 1:
                        client = _a.sent();
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 6, 8, 10]);
                        return [4 /*yield*/, client.query('BEGIN')];
                    case 3:
                        _a.sent();
                        return [4 /*yield*/, action({
                                createCommand: function () { return new PostgresBridgeToSQLCommand(client); },
                                entityIsUsedForTheFirstTime: this.entityIsUsedForTheFirstTime,
                                transaction: function () {
                                    throw 'nested transactions not allowed';
                                },
                                getLimitSqlSyntax: this.getLimitSqlSyntax,
                                supportsJsonColumnType: this.supportsJsonColumnType,
                                //@ts-ignore
                                pool: client,
                            })];
                    case 4:
                        _a.sent();
                        return [4 /*yield*/, client.query('COMMIT')];
                    case 5:
                        _a.sent();
                        return [3 /*break*/, 10];
                    case 6:
                        err_1 = _a.sent();
                        return [4 /*yield*/, client.query('ROLLBACK')];
                    case 7:
                        _a.sent();
                        throw err_1;
                    case 8: return [4 /*yield*/, client.release()];
                    case 9:
                        _a.sent();
                        return [7 /*endfinally*/];
                    case 10: return [2 /*return*/];
                }
            });
        });
    };
    return PostgresDataProvider;
}());
exports.PostgresDataProvider = PostgresDataProvider;
var PostgresBridgeToSQLCommand = /** @class */ (function () {
    function PostgresBridgeToSQLCommand(source) {
        this.source = source;
        this.values = [];
    }
    PostgresBridgeToSQLCommand.prototype.addParameterAndReturnSqlToken = function (val) {
        if (Array.isArray(val))
            val = JSON.stringify(val);
        this.values.push(val);
        return '$' + this.values.length;
    };
    PostgresBridgeToSQLCommand.prototype.execute = function (sql) {
        return this.source
            .query(sql, this.values)
            .then(function (r) { return new PostgresBridgeToSQLQueryResult(r); });
    };
    return PostgresBridgeToSQLCommand;
}());
var PostgresBridgeToSQLQueryResult = /** @class */ (function () {
    function PostgresBridgeToSQLQueryResult(r) {
        this.r = r;
        this.rows = r.rows;
    }
    PostgresBridgeToSQLQueryResult.prototype.getColumnKeyInResultForIndexInSelect = function (index) {
        return this.r.fields[index].name;
    };
    return PostgresBridgeToSQLQueryResult;
}());
function createPostgresConnection(options) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, createPostgresDataProvider(options)];
        });
    });
}
exports.createPostgresConnection = createPostgresConnection;
function createPostgresDataProvider(options) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var config, db, remult;
        return tslib_1.__generator(this, function (_a) {
            if (!options)
                options = {};
            config = {};
            if (options.configuration)
                if (options.configuration == 'heroku') {
                    config = {
                        connectionString: process.env.DATABASE_URL,
                        ssl: process.env.NODE_ENV !== 'production' && !options.sslInDev
                            ? false
                            : {
                                rejectUnauthorized: false,
                            },
                    };
                }
                else
                    config = options.configuration;
            else {
                if (!options.connectionString)
                    options.connectionString = process.env.DATABASE_URL;
            }
            if (!config.connectionString && options.connectionString) {
                config.connectionString = options.connectionString;
            }
            db = new sql_database_1.SqlDatabase(new PostgresDataProvider(new pg_1.Pool(config)));
            remult = new context_1.Remult();
            remult.dataProvider = db;
            return [2 /*return*/, db];
        });
    });
}
exports.createPostgresDataProvider = createPostgresDataProvider;
function preparePostgresQueueStorage(sql) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var c, JobsInQueueEntity, e;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    c = new context_1.Remult();
                    c.dataProvider = sql;
                    return [4 /*yield*/, Promise.resolve().then(function () { return tslib_1.__importStar(require('../server/expressBridge')); })];
                case 1:
                    JobsInQueueEntity = (_a.sent())
                        .JobsInQueueEntity;
                    e = c.repo(JobsInQueueEntity);
                    return [4 /*yield*/, sql.ensureSchema([e.metadata])];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, Promise.resolve().then(function () { return tslib_1.__importStar(require('../server/expressBridge')); })];
                case 3: return [2 /*return*/, new (_a.sent()).EntityQueueStorage(c.repo(JobsInQueueEntity))];
            }
        });
    });
}
exports.preparePostgresQueueStorage = preparePostgresQueueStorage;
