// remult/esm/src/data-providers/json-data-provider.js
import { ArrayEntityDataProvider } from './array-entity-data-provider';
export class JsonDataProvider {
    storage;
    constructor(storage) {
        this.storage = storage;
    }
    getEntityDataProvider(entity) {
        return new JsonEntityDataProvider(entity, this.storage);
    }
    async transaction(action) {
        await action(this);
    }
}
class JsonEntityDataProvider {
    entity;
    helper;
    constructor(entity, helper) {
        this.entity = entity;
        this.helper = helper;
    }
    async loadEntityData(what) {
        let data = [];
        let dbName = await this.entity.getDbName();
        let s = this.helper.getItem(dbName);
        if (s)
            data = JSON.parse(s);
        let dp = new ArrayEntityDataProvider(this.entity, data);
        return what(dp, () => this.helper.setItem(dbName, JSON.stringify(data, undefined, 2)));
    }
    p = Promise.resolve();
    find(options) {
        return (this.p = this.p.then(() => this.loadEntityData((dp, save) => dp.find(options))));
    }
    count(where) {
        return (this.p = this.p.then(() => this.loadEntityData((dp, save) => dp.count(where))));
    }
    update(id, data) {
        return (this.p = this.p.then(() => this.loadEntityData((dp, save) => dp.update(id, data).then((x) => {
            save();
            return x;
        }))));
    }
    delete(id) {
        return (this.p = this.p.then(() => this.loadEntityData((dp, save) => dp.delete(id).then((x) => {
            save();
            return x;
        }))));
    }
    async insert(data) {
        return (this.p = this.p.then(() => this.loadEntityData((dp, save) => dp.insert(data).then((x) => {
            save();
            return x;
        }))));
    }
}
