// remult/src/data-providers/array-entity-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayEntityDataProvider = void 0;
var tslib_1 = require("tslib");
var CompoundIdField_1 = require("../CompoundIdField");
var filter_interfaces_1 = require("../filter/filter-interfaces");
var ArrayEntityDataProvider = /** @class */ (function () {
    function ArrayEntityDataProvider(entity, rows) {
        var e_1, _a;
        this.entity = entity;
        this.rows = rows;
        if (!rows)
            rows = [];
        else {
            try {
                for (var rows_1 = tslib_1.__values(rows), rows_1_1 = rows_1.next(); !rows_1_1.done; rows_1_1 = rows_1.next()) {
                    var r = rows_1_1.value;
                    this.verifyThatRowHasAllNotNullColumns(r);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (rows_1_1 && !rows_1_1.done && (_a = rows_1.return)) _a.call(rows_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    }
    ArrayEntityDataProvider.rawFilter = function (filter) {
        var _a;
        return _a = {},
            _a[filter_interfaces_1.customDatabaseFilterToken] = {
                arrayFilter: filter,
            },
            _a;
    };
    ArrayEntityDataProvider.prototype.verifyThatRowHasAllNotNullColumns = function (r) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (!f.isServerExpression)
                    if (!f.allowNull) {
                        if (r[f.key] === undefined || r[f.key] === null) {
                            var val = undefined;
                            if (f.valueType === Boolean)
                                val = false;
                            else if (f.valueType === Number)
                                val = 0;
                            else if (f.valueType === String)
                                val = '';
                            r[f.key] = val;
                        }
                    }
                    else if (r[f.key] === undefined)
                        r[f.key] = null;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    ArrayEntityDataProvider.prototype.count = function (where) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rows, j, i, x;
            return tslib_1.__generator(this, function (_a) {
                rows = this.rows;
                j = 0;
                for (i = 0; i < rows.length; i++) {
                    if (!where) {
                        j++;
                    }
                    else {
                        x = new FilterConsumerBridgeToObject(rows[i]);
                        where.__applyToConsumer(x);
                        if (x.ok)
                            j++;
                    }
                }
                return [2 /*return*/, j];
            });
        });
    };
    ArrayEntityDataProvider.prototype.find = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rows;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                rows = this.rows;
                if (options) {
                    if (options.where) {
                        rows = rows.filter(function (i) {
                            var x = new FilterConsumerBridgeToObject(i);
                            options.where.__applyToConsumer(x);
                            return x.ok;
                        });
                    }
                    if (options.orderBy) {
                        rows = rows.sort(function (a, b) {
                            return options.orderBy.compare(a, b);
                        });
                    }
                    rows = pageArray(rows, options);
                }
                if (rows)
                    return [2 /*return*/, rows.map(function (i) {
                            return _this.translateFromJson(i);
                        })];
                return [2 /*return*/];
            });
        });
    };
    ArrayEntityDataProvider.prototype.translateFromJson = function (row) {
        var e_3, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                result[col.key] = col.valueConverter.fromJson(row[col.key]);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return result;
    };
    ArrayEntityDataProvider.prototype.translateToJson = function (row) {
        var e_4, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                result[col.key] = col.valueConverter.toJson(row[col.key]);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return result;
    };
    ArrayEntityDataProvider.prototype.idMatches = function (id) {
        var _this = this;
        return function (item) {
            var x = new FilterConsumerBridgeToObject(item);
            filter_interfaces_1.Filter.fromEntityFilter(_this.entity, _this.entity.idMetadata.getIdFilter(id)).__applyToConsumer(x);
            return x.ok;
        };
    };
    ArrayEntityDataProvider.prototype.update = function (id, data) {
        var e_5, _a;
        var idMatches = this.idMatches(id);
        var keys = Object.keys(data);
        for (var i = 0; i < this.rows.length; i++) {
            var r = this.rows[i];
            if (idMatches(r)) {
                var newR = tslib_1.__assign({}, r);
                try {
                    for (var _b = (e_5 = void 0, tslib_1.__values(this.entity.fields)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (!f.dbReadOnly && !f.isServerExpression) {
                            if (keys.includes(f.key)) {
                                newR[f.key] = f.valueConverter.toJson(data[f.key]);
                            }
                        }
                    }
                }
                catch (e_5_1) { e_5 = { error: e_5_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_5) throw e_5.error; }
                }
                this.verifyThatRowHasAllNotNullColumns(newR);
                this.rows[i] = newR;
                return Promise.resolve(this.translateFromJson(this.rows[i]));
            }
        }
        throw new Error("couldn't find id to update: " + id);
    };
    ArrayEntityDataProvider.prototype.delete = function (id) {
        var idMatches = this.idMatches(id);
        for (var i = 0; i < this.rows.length; i++) {
            if (idMatches(this.rows[i])) {
                this.rows.splice(i, 1);
                return Promise.resolve();
            }
        }
        throw new Error("couldn't find id to delete: " + id);
    };
    ArrayEntityDataProvider.prototype.insert = function (data) {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var idf, _b, _c, row, j;
            var e_6, _d;
            return tslib_1.__generator(this, function (_e) {
                idf = this.entity.idMetadata.field;
                if (!(idf instanceof CompoundIdField_1.CompoundIdField)) {
                    if (((_a = idf.options.valueConverter) === null || _a === void 0 ? void 0 : _a.fieldTypeInDb) === 'autoincrement') {
                        data[idf.key] = 1;
                        try {
                            for (_b = tslib_1.__values(this.rows), _c = _b.next(); !_c.done; _c = _b.next()) {
                                row = _c.value;
                                if (row[idf.key] >= data[idf.key])
                                    data[idf.key] = row[idf.key] + 1;
                            }
                        }
                        catch (e_6_1) { e_6 = { error: e_6_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_d = _b.return)) _d.call(_b);
                            }
                            finally { if (e_6) throw e_6.error; }
                        }
                    }
                    if (data[idf.key])
                        this.rows.forEach(function (i) {
                            if (data[idf.key] == i[idf.key])
                                throw Error('id already exists');
                        });
                }
                j = this.translateToJson(data);
                this.verifyThatRowHasAllNotNullColumns(j);
                this.rows.push(j);
                return [2 /*return*/, Promise.resolve(this.translateFromJson(j))];
            });
        });
    };
    return ArrayEntityDataProvider;
}());
exports.ArrayEntityDataProvider = ArrayEntityDataProvider;
function pageArray(rows, options) {
    if (!options)
        return rows;
    if (!options.limit)
        return rows;
    var page = 1;
    if (options.page)
        page = options.page;
    if (page < 1)
        page = 1;
    var x = 0;
    return rows.filter(function (i) {
        x++;
        var max = page * options.limit;
        var min = max - options.limit;
        return x > min && x <= max;
    });
}
var FilterConsumerBridgeToObject = /** @class */ (function () {
    function FilterConsumerBridgeToObject(row) {
        this.row = row;
        this.ok = true;
    }
    FilterConsumerBridgeToObject.prototype.databaseCustom = function (databaseCustom) {
        if (databaseCustom && databaseCustom.arrayFilter) {
            if (!databaseCustom.arrayFilter(this.row))
                this.ok = false;
        }
    };
    FilterConsumerBridgeToObject.prototype.custom = function (key, customItem) {
        throw new Error('Custom Filter should be translated before it gets here');
    };
    FilterConsumerBridgeToObject.prototype.or = function (orElements) {
        var e_7, _a;
        try {
            for (var orElements_1 = tslib_1.__values(orElements), orElements_1_1 = orElements_1.next(); !orElements_1_1.done; orElements_1_1 = orElements_1.next()) {
                var element = orElements_1_1.value;
                var filter = new FilterConsumerBridgeToObject(this.row);
                element.__applyToConsumer(filter);
                if (filter.ok) {
                    return;
                }
            }
        }
        catch (e_7_1) { e_7 = { error: e_7_1 }; }
        finally {
            try {
                if (orElements_1_1 && !orElements_1_1.done && (_a = orElements_1.return)) _a.call(orElements_1);
            }
            finally { if (e_7) throw e_7.error; }
        }
        this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isNull = function (col) {
        if (this.row[col.key] != null)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isNotNull = function (col) {
        if (this.row[col.key] == null)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isIn = function (col, val) {
        var e_8, _a;
        try {
            for (var val_1 = tslib_1.__values(val), val_1_1 = val_1.next(); !val_1_1.done; val_1_1 = val_1.next()) {
                var v = val_1_1.value;
                if (this.row[col.key] == col.valueConverter.toJson(v)) {
                    return;
                }
            }
        }
        catch (e_8_1) { e_8 = { error: e_8_1 }; }
        finally {
            try {
                if (val_1_1 && !val_1_1.done && (_a = val_1.return)) _a.call(val_1);
            }
            finally { if (e_8) throw e_8.error; }
        }
        this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isEqualTo = function (col, val) {
        if (this.row[col.key] != col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isDifferentFrom = function (col, val) {
        if (this.row[col.key] == col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isGreaterOrEqualTo = function (col, val) {
        if (this.row[col.key] < col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isGreaterThan = function (col, val) {
        if (this.row[col.key] <= col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isLessOrEqualTo = function (col, val) {
        if (this.row[col.key] > col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isLessThan = function (col, val) {
        if (this.row[col.key] >= col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.containsCaseInsensitive = function (col, val) {
        var v = this.row[col.key];
        if (!v) {
            this.ok = false;
            return;
        }
        var s = '' + v;
        if (val)
            val = col.valueConverter.toJson(val);
        if (val)
            val = val.toString().toLowerCase();
        if (s.toLowerCase().indexOf(val) < 0)
            this.ok = false;
    };
    return FilterConsumerBridgeToObject;
}());
