// remult/src/data-providers/json-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonDataProvider = void 0;
var tslib_1 = require("tslib");
var array_entity_data_provider_1 = require("./array-entity-data-provider");
var JsonDataProvider = /** @class */ (function () {
    function JsonDataProvider(storage) {
        this.storage = storage;
    }
    JsonDataProvider.prototype.getEntityDataProvider = function (entity) {
        return new JsonEntityDataProvider(entity, this.storage);
    };
    JsonDataProvider.prototype.transaction = function (action) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, action(this)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return JsonDataProvider;
}());
exports.JsonDataProvider = JsonDataProvider;
var JsonEntityDataProvider = /** @class */ (function () {
    function JsonEntityDataProvider(entity, helper) {
        this.entity = entity;
        this.helper = helper;
        this.p = Promise.resolve();
    }
    JsonEntityDataProvider.prototype.loadEntityData = function (what) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data, dbName, s, dp;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        data = [];
                        return [4 /*yield*/, this.entity.getDbName()];
                    case 1:
                        dbName = _a.sent();
                        s = this.helper.getItem(dbName);
                        if (s)
                            data = JSON.parse(s);
                        dp = new array_entity_data_provider_1.ArrayEntityDataProvider(this.entity, data);
                        return [2 /*return*/, what(dp, function () {
                                return _this.helper.setItem(dbName, JSON.stringify(data, undefined, 2));
                            })];
                }
            });
        });
    };
    JsonEntityDataProvider.prototype.find = function (options) {
        var _this = this;
        return (this.p = this.p.then(function () {
            return _this.loadEntityData(function (dp, save) { return dp.find(options); });
        }));
    };
    JsonEntityDataProvider.prototype.count = function (where) {
        var _this = this;
        return (this.p = this.p.then(function () {
            return _this.loadEntityData(function (dp, save) { return dp.count(where); });
        }));
    };
    JsonEntityDataProvider.prototype.update = function (id, data) {
        var _this = this;
        return (this.p = this.p.then(function () {
            return _this.loadEntityData(function (dp, save) {
                return dp.update(id, data).then(function (x) {
                    save();
                    return x;
                });
            });
        }));
    };
    JsonEntityDataProvider.prototype.delete = function (id) {
        var _this = this;
        return (this.p = this.p.then(function () {
            return _this.loadEntityData(function (dp, save) {
                return dp.delete(id).then(function (x) {
                    save();
                    return x;
                });
            });
        }));
    };
    JsonEntityDataProvider.prototype.insert = function (data) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, (this.p = this.p.then(function () {
                        return _this.loadEntityData(function (dp, save) {
                            return dp.insert(data).then(function (x) {
                                save();
                                return x;
                            });
                        });
                    }))];
            });
        });
    };
    return JsonEntityDataProvider;
}());
