// remult/esm/src/remult3/getEntityRef.js
export function getEntityRef(entity, throwException = true) {
    let x = entity[entityMember];
    if (!x && throwException)
        throw new Error('item ' +
            entity.constructor.name +
            ' was not initialized using a context');
    return x;
}
export const entityMember = Symbol('entityMember');
export const entityInfo = Symbol('entityInfo');
export const entityInfo_key = Symbol('entityInfo_key');
export function getEntitySettings(entity, throwError = true) {
    if (entity === undefined)
        if (throwError) {
            throw new Error('Undefined is not an entity :)');
        }
        else
            return undefined;
    let info = Reflect.getMetadata(entityInfo, entity);
    if (!info && throwError)
        throw new Error(entity.prototype.constructor.name +
            " is not a known entity, did you forget to set @Entity() or did you forget to add the '@' before the call to Entity?");
    return info;
}
export function getEntityKey(entity) {
    return Reflect.getMetadata(entityInfo_key, entity);
}
