// remult/esm/src/filter/filter-consumer-bridge-to-sql-request.js
import { SqlDatabase } from '../data-providers/sql-database';
import { getEntityMetadata } from '../remult3/RepositoryImplementation';
export class FilterConsumerBridgeToSqlRequest {
    r;
    nameProvider;
    where = '';
    _addWhere = true;
    promises = [];
    async resolveWhere() {
        while (this.promises.length > 0) {
            let p = this.promises;
            this.promises = [];
            for (const pr of p) {
                await pr;
            }
        }
        return this.where;
    }
    constructor(r, nameProvider) {
        this.r = r;
        this.nameProvider = nameProvider;
    }
    custom(key, customItem) {
        throw new Error('Custom filter should be translated before it gets here');
    }
    or(orElements) {
        let statement = '';
        this.promises.push((async () => {
            for (const element of orElements) {
                let f = new FilterConsumerBridgeToSqlRequest(this.r, this.nameProvider);
                f._addWhere = false;
                element.__applyToConsumer(f);
                let where = await f.resolveWhere();
                if (!where)
                    return; //since if any member of or is empty, then the entire or is irrelevant
                if (where.length > 0) {
                    if (statement.length > 0) {
                        statement += ' or ';
                    }
                    if (orElements.length > 1) {
                        statement += '(' + where + ')';
                    }
                    else
                        statement += where;
                }
            }
            this.addToWhere('(' + statement + ')');
        })());
    }
    isNull(col) {
        this.promises.push((async () => this.addToWhere(this.nameProvider.$dbNameOf(col) + ' is null'))());
    }
    isNotNull(col) {
        this.promises.push((async () => this.addToWhere(this.nameProvider.$dbNameOf(col) + ' is not null'))());
    }
    isIn(col, val) {
        this.promises.push((async () => {
            if (val && val.length > 0)
                this.addToWhere(this.nameProvider.$dbNameOf(col) +
                    ' in (' +
                    val
                        .map((x) => this.r.addParameterAndReturnSqlToken(col.valueConverter.toDb(x)))
                        .join(',') +
                    ')');
            else
                this.addToWhere('1 = 0 /*isIn with no values*/');
        })());
    }
    isEqualTo(col, val) {
        this.add(col, val, '=');
    }
    isDifferentFrom(col, val) {
        this.add(col, val, '<>');
    }
    isGreaterOrEqualTo(col, val) {
        this.add(col, val, '>=');
    }
    isGreaterThan(col, val) {
        this.add(col, val, '>');
    }
    isLessOrEqualTo(col, val) {
        this.add(col, val, '<=');
    }
    isLessThan(col, val) {
        this.add(col, val, '<');
    }
    containsCaseInsensitive(col, val) {
        this.promises.push((async () => {
            this.addToWhere('lower (' +
                this.nameProvider.$dbNameOf(col) +
                ") like lower ('%" +
                val.replace(/'/g, "''") +
                "%')");
        })());
    }
    add(col, val, operator) {
        this.promises.push((async () => {
            let x = this.nameProvider.$dbNameOf(col) +
                ' ' +
                operator +
                ' ' +
                this.r.addParameterAndReturnSqlToken(col.valueConverter.toDb(val));
            this.addToWhere(x);
        })());
    }
    addToWhere(x) {
        if (this.where.length == 0) {
            if (this._addWhere)
                this.where += ' where ';
        }
        else
            this.where += ' and ';
        this.where += x;
    }
    databaseCustom(databaseCustom) {
        this.promises.push((async () => {
            if (databaseCustom?.buildSql) {
                let item = new CustomSqlFilterBuilder(this.r);
                await databaseCustom.buildSql(item);
                if (item.sql) {
                    this.addToWhere('(' + item.sql + ')');
                }
            }
        })());
    }
}
export class CustomSqlFilterBuilder {
    r;
    constructor(r) {
        this.r = r;
    }
    sql = '';
    addParameterAndReturnSqlToken(val, field) {
        if (field)
            val = field.valueConverter.toDb(val);
        return this.r.addParameterAndReturnSqlToken(val);
    }
    async filterToRaw(repo, condition) {
        return SqlDatabase.filterToRaw(repo, condition, this);
    }
}
export function isDbReadonly(field, dbNames) {
    return (field.dbReadOnly ||
        field.isServerExpression ||
        dbNames.$dbNameOf(field) != field.options.dbName);
}
export async function dbNamesOf(repo) {
    var meta = getEntityMetadata(repo);
    const result = {
        $entityName: await meta.getDbName(),
        toString: () => result.$entityName,
        $dbNameOf: (field) => {
            var key;
            if (typeof field === 'string')
                key = field;
            else
                key = field.key;
            return result[key];
        },
    };
    for (const field of meta.fields) {
        result[field.key] = await field.getDbName();
    }
    return result;
}
