// remult/server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRemultServer = exports.JsonFileDataProvider = exports.JsonEntityFileStorage = exports.DataProviderLiveQueryStorage = exports.SseSubscriptionServer = void 0;
var tslib_1 = require("tslib");
var expressBridge_1 = require("./expressBridge");
var initAsyncHooks_1 = require("./initAsyncHooks");
var SseSubscriptionServer_1 = require("../SseSubscriptionServer");
Object.defineProperty(exports, "SseSubscriptionServer", { enumerable: true, get: function () { return SseSubscriptionServer_1.SseSubscriptionServer; } });
var data_provider_live_query_storage_1 = require("../live-query/data-provider-live-query-storage");
Object.defineProperty(exports, "DataProviderLiveQueryStorage", { enumerable: true, get: function () { return data_provider_live_query_storage_1.DataProviderLiveQueryStorage; } });
var JsonEntityFileStorage_1 = require("./JsonEntityFileStorage");
Object.defineProperty(exports, "JsonEntityFileStorage", { enumerable: true, get: function () { return JsonEntityFileStorage_1.JsonEntityFileStorage; } });
Object.defineProperty(exports, "JsonFileDataProvider", { enumerable: true, get: function () { return JsonEntityFileStorage_1.JsonFileDataProvider; } });
function createRemultServer(options, serverCoreOptions) {
    var _this = this;
    (0, initAsyncHooks_1.initAsyncHooks)();
    return (0, expressBridge_1.createRemultServerCore)(options, serverCoreOptions || {
        buildGenericRequestInfo: function (req) { return req; },
        getRequestBody: function (req) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, req.body];
        }); }); },
    });
}
exports.createRemultServer = createRemultServer;
