// remult/esm/src/column.js
import { assign } from '../assign';
import { __updateEntityBasedOnWhere } from './remult3/__updateEntityBasedOnWhere';
import { getEntityRef } from './remult3/getEntityRef';
export function makeTitle(name) {
    // insert a space before all caps
    return (name
        .replace(/([A-Z])/g, ' $1')
        // uppercase the first character
        .replace(/^./, (str) => str.toUpperCase())
        .replace('Email', 'eMail')
        .replace(' I D', ' ID'));
}
export class LookupColumn {
    repository;
    id;
    toJson() {
        if (this.item === null)
            return null;
        return this.repository.toJson(this.item);
    }
    setId(val) {
        if (this.repository.metadata.idMetadata.field.valueType == Number)
            val = +val;
        this.id = val;
    }
    waitLoadOf(id) {
        if (id === undefined || id === null)
            return null;
        return this.repository.getCachedByIdAsync(id);
    }
    get(id) {
        if (id === undefined || id === null)
            return null;
        return this.repository.getCachedById(id);
    }
    storedItem;
    set(item) {
        this.storedItem = undefined;
        if (item) {
            if (typeof item === 'string' || typeof item === 'number')
                this.id = item;
            else {
                let eo = getEntityRef(item, false);
                if (eo) {
                    this.repository.addToCache(item);
                    this.id = eo.getId();
                }
                else {
                    this.storedItem = { item };
                    this.id = item[this.repository.metadata.idMetadata.field.key];
                }
            }
        }
        else if (item === null) {
            this.id = null;
        }
        else {
            this.id = undefined;
        }
    }
    constructor(repository, id) {
        this.repository = repository;
        this.id = id;
    }
    get item() {
        if (this.storedItem)
            return this.storedItem.item;
        return this.get(this.id);
    }
    async waitLoad() {
        return this.waitLoadOf(this.id);
    }
}
9000;
export class OneToMany {
    provider;
    settings;
    constructor(provider, settings) {
        this.provider = provider;
        this.settings = settings;
    }
    _items;
    _currentPromise;
    get lazyItems() {
        this.load();
        return this._items;
    }
    async load() {
        if (this._currentPromise != null)
            return this._currentPromise;
        if (this._items === undefined)
            this._items = [];
        return (this._currentPromise = this.find().then((x) => {
            this._items.splice(0);
            this._items.push(...x);
            return this._items;
        }));
    }
    async find() {
        return this.provider.find(this.settings);
    }
    create(item) {
        let r = this.provider.create();
        __updateEntityBasedOnWhere(this.provider.metadata, this.settings.where, r);
        assign(r, item);
        return r;
    }
}
