// remult/src/column.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneToMany = exports.LookupColumn = exports.makeTitle = void 0;
var tslib_1 = require("tslib");
var assign_1 = require("../assign");
var __updateEntityBasedOnWhere_1 = require("./remult3/__updateEntityBasedOnWhere");
var getEntityRef_1 = require("./remult3/getEntityRef");
function makeTitle(name) {
    // insert a space before all caps
    return (name
        .replace(/([A-Z])/g, ' $1')
        // uppercase the first character
        .replace(/^./, function (str) { return str.toUpperCase(); })
        .replace('Email', 'eMail')
        .replace(' I D', ' ID'));
}
exports.makeTitle = makeTitle;
var LookupColumn = /** @class */ (function () {
    function LookupColumn(repository, id) {
        this.repository = repository;
        this.id = id;
    }
    LookupColumn.prototype.toJson = function () {
        if (this.item === null)
            return null;
        return this.repository.toJson(this.item);
    };
    LookupColumn.prototype.setId = function (val) {
        if (this.repository.metadata.idMetadata.field.valueType == Number)
            val = +val;
        this.id = val;
    };
    LookupColumn.prototype.waitLoadOf = function (id) {
        if (id === undefined || id === null)
            return null;
        return this.repository.getCachedByIdAsync(id);
    };
    LookupColumn.prototype.get = function (id) {
        if (id === undefined || id === null)
            return null;
        return this.repository.getCachedById(id);
    };
    LookupColumn.prototype.set = function (item) {
        this.storedItem = undefined;
        if (item) {
            if (typeof item === 'string' || typeof item === 'number')
                this.id = item;
            else {
                var eo = (0, getEntityRef_1.getEntityRef)(item, false);
                if (eo) {
                    this.repository.addToCache(item);
                    this.id = eo.getId();
                }
                else {
                    this.storedItem = { item: item };
                    this.id = item[this.repository.metadata.idMetadata.field.key];
                }
            }
        }
        else if (item === null) {
            this.id = null;
        }
        else {
            this.id = undefined;
        }
    };
    Object.defineProperty(LookupColumn.prototype, "item", {
        get: function () {
            if (this.storedItem)
                return this.storedItem.item;
            return this.get(this.id);
        },
        enumerable: false,
        configurable: true
    });
    LookupColumn.prototype.waitLoad = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, this.waitLoadOf(this.id)];
            });
        });
    };
    return LookupColumn;
}());
exports.LookupColumn = LookupColumn;
9000;
var OneToMany = /** @class */ (function () {
    function OneToMany(provider, settings) {
        this.provider = provider;
        this.settings = settings;
    }
    Object.defineProperty(OneToMany.prototype, "lazyItems", {
        get: function () {
            this.load();
            return this._items;
        },
        enumerable: false,
        configurable: true
    });
    OneToMany.prototype.load = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                if (this._currentPromise != null)
                    return [2 /*return*/, this._currentPromise];
                if (this._items === undefined)
                    this._items = [];
                return [2 /*return*/, (this._currentPromise = this.find().then(function (x) {
                        var _a;
                        _this._items.splice(0);
                        (_a = _this._items).push.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(x), false));
                        return _this._items;
                    }))];
            });
        });
    };
    OneToMany.prototype.find = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, this.provider.find(this.settings)];
            });
        });
    };
    OneToMany.prototype.create = function (item) {
        var r = this.provider.create();
        (0, __updateEntityBasedOnWhere_1.__updateEntityBasedOnWhere)(this.provider.metadata, this.settings.where, r);
        (0, assign_1.assign)(r, item);
        return r;
    };
    return OneToMany;
}());
exports.OneToMany = OneToMany;
