// remult/src/remult-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.remult = exports.RemultProxy = void 0;
var tslib_1 = require("tslib");
var defaultRemult;
function defaultFactory() {
    if (!defaultRemult) {
        defaultRemult = RemultProxy.defaultRemultFactory();
    }
    return defaultRemult;
}
/*@internal*/
var RemultProxy = /** @class */ (function () {
    function RemultProxy() {
        var _this = this;
        /*@internal*/
        this.remultFactory = function () { return defaultFactory(); };
        this.repoCache = new Map();
        //@ts-ignore
        this.repo = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var self = _this;
            var entityCache = self.repoCache.get(args[0]);
            if (!entityCache) {
                self.repoCache.set(args[0], (entityCache = new Map()));
            }
            var result = entityCache.get(args[1]);
            if (result)
                return result;
            result = {
                get fields() {
                    var _a;
                    return (_a = self.remultFactory()).repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).metadata.fields;
                },
                validate: function (a, b) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).validate(a, b);
                },
                addEventListener: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .addEventListener.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                count: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .count.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                create: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .create.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                delete: function (args2) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).delete(args2);
                },
                find: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .find.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                findFirst: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .findFirst.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                findId: function (a, b) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).findId(a, b);
                },
                //@ts-ignore
                toJson: function (json) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).toJson(json);
                },
                fromJson: function (item, isNew) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).fromJson(item, isNew);
                },
                getEntityRef: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .getEntityRef.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                insert: function (args2) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).insert(args2);
                },
                liveQuery: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .liveQuery.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                //@ts-ignore
                addToCache: function (a) {
                    var _a;
                    return (_a = self.remultFactory()).repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).addToCache(a);
                },
                get metadata() {
                    var _a;
                    return (_a = self.remultFactory()).repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).metadata;
                },
                query: function () {
                    var _a, _b;
                    var args2 = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args2[_i] = arguments[_i];
                    }
                    return (_a = (_b = self
                        .remultFactory())
                        .repo.apply(_b, tslib_1.__spreadArray([], tslib_1.__read(args), false)))
                        .query.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args2), false));
                },
                save: function (args2) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).save(args2);
                },
                update: function (a, b) {
                    var _a;
                    return (_a = self
                        .remultFactory())
                        .repo.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(args), false)).update(a, b);
                },
            };
            entityCache.set(args[1], result);
            return result;
        };
    }
    Object.defineProperty(RemultProxy.prototype, "liveQuerySubscriber", {
        /* @internal*/
        get: function () {
            return this.remultFactory().liveQuerySubscriber;
        },
        /* @internal*/
        set: function (val) {
            this.remultFactory().liveQuerySubscriber = val;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "liveQueryStorage", {
        /* @internal*/
        get: function () {
            return this.remultFactory().liveQueryStorage;
        },
        /* @internal*/
        set: function (val) {
            this.remultFactory().liveQueryStorage = val;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "liveQueryPublisher", {
        /* @internal*/
        get: function () {
            return this.remultFactory().liveQueryPublisher;
        },
        /* @internal*/
        set: function (val) {
            this.remultFactory().liveQueryPublisher = val;
        },
        enumerable: false,
        configurable: true
    });
    RemultProxy.prototype.call = function (backendMethod, self) {
        var _a;
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        return (_a = this.remultFactory()).call.apply(_a, tslib_1.__spreadArray([backendMethod, self], tslib_1.__read(args), false));
    };
    Object.defineProperty(RemultProxy.prototype, "context", {
        get: function () {
            return this.remultFactory().context;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "dataProvider", {
        get: function () {
            return this.remultFactory().dataProvider;
        },
        set: function (provider) {
            this.remultFactory().dataProvider = provider;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "repCache", {
        /*@internal*/
        get: function () {
            return this.remultFactory().repCache;
        },
        enumerable: false,
        configurable: true
    });
    RemultProxy.prototype.authenticated = function () {
        return this.remultFactory().authenticated();
    };
    RemultProxy.prototype.isAllowed = function (roles) {
        return this.remultFactory().isAllowed(roles);
    };
    RemultProxy.prototype.isAllowedForInstance = function (instance, allowed) {
        return this.remultFactory().isAllowedForInstance(instance, allowed);
    };
    RemultProxy.prototype.clearAllCache = function () {
        return this.remultFactory().clearAllCache();
    };
    /*@internal*/
    RemultProxy.prototype.resetFactory = function () {
        this.remultFactory = function () { return defaultFactory(); };
    };
    Object.defineProperty(RemultProxy.prototype, "user", {
        get: function () {
            return this.remultFactory().user;
        },
        set: function (info) {
            this.remultFactory().user = info;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "apiClient", {
        get: function () {
            return this.remultFactory().apiClient;
        },
        set: function (client) {
            this.remultFactory().apiClient = client;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RemultProxy.prototype, "subscriptionServer", {
        get: function () {
            return this.remultFactory().subscriptionServer;
        },
        set: function (value) {
            this.remultFactory().subscriptionServer = value;
        },
        enumerable: false,
        configurable: true
    });
    return RemultProxy;
}());
exports.RemultProxy = RemultProxy;
exports.remult = new RemultProxy();
